src/components/questionStructure/Task/variants/HandwritingTask/Canvas.tsx

```diff
@@ -0,0 +1,138 @@
+  onAnswerChange,
+}) => {
+  const { updateStrokes } = useLiveUpdates(username, onAnswerChange)
+  const [excalidrawAPI, setExcalidrawAPI] = useState<ExcalidrawImperativeAPI | null>(null)
+  const [clearDialogOpen, setClearDialogOpen] = useState(false)
+
+  const clearCanvas = useCallback(() => {
+    updateStrokes({ elements: [] })
+    excalidrawAPI?.updateScene({ elements: [] })
+  }, [excalidrawAPI, updateStrokes])
+
+  // HACK: Disable Excalidraw right click menu
+  const excalidrawWrapperRef = useRef<HTMLDivElement | null>(null)
+  useEffect(() => {
+    if (!excalidrawAPI || !excalidrawWrapperRef.current) return
+
+    const canvas = excalidrawWrapperRef.current.getElementsByClassName('interactive')[0]
+    canvas?.addEventListener('contextmenu', stopEvent)
+
+    // Force a canvas resize when the dialog opens
+    setTimeout(() => window.dispatchEvent(new Event('resize')), 100)
+
+    return () => canvas?.removeEventListener('contextmenu', stopEvent)
+  }, [excalidrawAPI, excalidrawWrapperRef]) // wait for Excalidraw to load
+
+  // Excalidraw API handlers
+  useEffect(() => {
+    // Call updateStrokes whenever the user finishes drawing/erasing/moving items
+    const pointerUpHandler = ({ type }: AppState['activeTool']): void => {
+      if (type === 'freedraw' || type === 'eraser' || type === 'selection') {
+        setTimeout(() => {
+          const elements = excalidrawAPI!.getSceneElements()!
+          updateStrokes({ elements })
+        })
+      }
+    }
+
+    const pointerUpCleanup = excalidrawAPI?.onPointerUp(pointerUpHandler)
+
+    return () => {
+      if (pointerUpCleanup) pointerUpCleanup()
+    }
+  }, [excalidrawAPI, updateStrokes])
+
+  // Disable pasting of text
+  const pasteHandler = (data: ClipboardData): boolean => !data.text
+
+  // Only allow specific keyboard actions
+  const keyDownHandler: KeyboardEventHandler<HTMLDivElement> = useCallback(
+    (event) => {
+      // Prevent writing in text boxes
+      if (document.activeElement?.nodeName === 'TEXTAREA') {
+        stopEvent(event)
+      }
+
+      // Only allow certain keys to be pressed in the canvas
+      // This prevents access to hidden tools e.g. pressing "r" to enter rectangle mode
+      if (
+        event.code === 'Backspace' || // Allow backspace for deletions
+        event.code.includes('Arrow') || // Allow arrow keys for moving elements
+        ((event.ctrlKey || event.metaKey) && // Allow specific keyboard shortcuts
+          ['KeyC', 'KeyV', 'KeyZ', 'KeyY', 'Equal', 'Minus', 'Digit0'].includes(event.code))
+      ) {
+        // Wait for the key event to be processed before updating the strokes
+        setTimeout(() => {
+          updateStrokes({ elements: excalidrawAPI?.getSceneElements() })
+        })
+      } else if (
+        // Excalidraw shortcuts for tools that should not be blocked
+        !['KeyH', 'KeyE', 'KeyV', 'KeyP', 'Digit1', 'Digit0', 'Digit7'].includes(event.code)
+      ) {
+        stopEvent(event)
+      }
```

Comment:
How about extracting the list into a constant? Like `ALLOWED_TOOL_SHORTCUTS`. Also helps with conciseness

```suggestion
      } else if (!ALLOWED_TOOL_SHORTCUTS.includes(event.code)) stopEvent(event)
```

Reply:
Extracted in 58c0c643c7f46255c5d92023456f56e5ce29e920